### DISCUSSION TITLE
Pampac

### DISCUSSION THREAD

**Discussion by joancf** [GateNLP/python-gatenlp#129] [Help Q&A] [answered]
One of the reasons to use GateNlp was Pampac. I find powerful the capacity to merge different kinds of annotations in the matching.

Now I'm facing a "simple" problem: Detec sentences that start with a Number and  a :   like 
" 10. 

I tried with:
```
r2 = Rule(
        Seq(AnnAt("Sentence"),Ann("Token", features=dict(is_digit=True)), Ann("Token", features=dict(lemma=".")) , name="seq2"),
        AddAnn(name="seq2", type="RefNum")
)
outset = doc.annset("Out2")
annSet=doc.annset("Original markups").with_type(["Paragraph","Sentence","Token"])
pars = annSet.with_type("Paragraph")
for par in pars:
        anns= annSet.within(par)
        Pampac(r2).run(doc, anns, outset=outset)
doc

```
But does not detect nothing  (if I remove AnnAt("Sentence") then I get some matches (some of them not at the begining)

Thanks!!

**Comment by johann-petrak** [comment, marked as answer]
Could you please describe more exactly the pattern you intend to match? 
I notice that the rule you try will look for, in order, a Sentence annotation, then Token annotation with the is_digit feature then a Token annotation withe lemma="." feature. But I think you want to match a Token that starts together with a sentence and has the is_digit feature true followed by another token etc. 

In order to match an annotation at the beginning of a sentence you need a construct like 
```python
Ann("Token", features=dict(is_digit=True)).at("Sentence")
```

Note that the difference between `Ann` and `AnnAt` relates to something different: the annotations in the input set form a specific sequence, and even if e.g. Ann1 and Ann2 start at the same offset, the will occur in a certain order in the input annotation set. `Ann` will match the next annotation in that sequence, while `AnnAt` will match any annotation at the next offset where one or more annotations occur in the sequence. 

BTW it is usually a good idea to restrict the input set to only annotations that are actually ever matched in a rule.

In your case something like this may work:
```python
r2 = Rule(
        Seq(
            AnnAt("Token", features=dict(is_digit=True)).at("Sentence"), 
            Ann("Token", features=dict(lemma=".")) , name="seq2"),
        AddAnn(name="seq2", type="RefNum")
)
```

**Reply by joancf** [reply]
Thanks Johann, 
It worked but even knowing the solution I don't know how to get there using the documentation ;-( . Ithink that each function needs an  example.. (and maybe the presentation or the notebook need also more examples) at least each functionality should appear in one example. I would help writing them, but I don't have the knowledge to do so.

**Reply by johann-petrak** [reply]
Yes, you are right, currently a lot of what can be done is only documented in the PythonDoc of the code or through the code itself. We definitely need more examples and a more detailled documentation. 
I am sorry, as we share this software as open source without explicit funding there is sometimes not enough time/resources to do this quickly.

**Reply by joancf** [reply]
I thank you for releasing it as open source, and documentation is always a problem!!  You don't need to apologize for that!
But better documentation -> more users ->  more feedback -> more bugs fixed ;-)
